Support Inkscape 1.0

fstokke opened the pull request:
Patch to make svg2shenzhen work with Inkscape 1.0 without warnings about deprecated APIs.

I have followed the [Updating your Extension for 1.0](https://wiki.inkscape.org/wiki/index.php/Updating_your_Extension_for_1.0) guidelines from Inkscape.

- Replaced OptionParser.add_option with the new arg_parser.add_argument. Changed type of dpi option from float to int to avoid error message when the option was later passed to the Inkscape executable.

- Removed use of deprected simplepath, simpletransform, simplestyle, cubicsuperpath and cspsubdiv packages. I have tried to use the replacements suggested in the deprecation warnings.

- Changed reading/writing of options.pickle. The options object in Inkscape 1.0 seems to contain a stream reader object that cannot be serialized. Replaced with export_options.pickle containing only the subset of the option fields that are actually used. Changed the filename to avoid reading existing incompatible options.debug in the .svg2shenzhen-cache directory.

- Replaced process.wait() with process.communicate() to avoid possible deadlock situation and warnings about unclosed files.

- Replaced Inkscape -e option with -o. -e does not seem to be supported in Inkscape 1.0

- Replaced bitmap2component.exe for Windows with the one in the 0.2.17 release zip. The one in the master branch seems to put the F.SilkS layer onto the F.Cu layer (see Issue #75). This problem also seems to be present in bitmap2components for Linux (have not tested for Mac OS).

I have done some light testing with some of my old svg2shenzhen projects. Seems to work OK, but more testing would probably be good.

Note that the changes will not be backwards compatible with Inkscape 0.9.x. People that need to use the old version of Inkscape will have to use old 0.2.17 version of svg2shenzhen.

fstokke committed:
Support Inkscape 1.0

okelk commented:
Cool. Tested with Inkscape 1.0 and the myfirstPCB.svg form the exaples folder.
It works and the PCB looks fine.

I get following messages:

C:\Program Files\Inkscape\lib\python3.8\subprocess.py:942: ResourceWarning: subprocess 24736 is still running
  _warn("subprocess %s is still running" % self.pid,
ResourceWarning: Enable tracemalloc to get the object allocation traceback
export.py:488: ResourceWarning: unclosed file <_io.BufferedReader name=4>
  self.openKicad(kicad_pcb_path)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
export.py:488: ResourceWarning: unclosed file <_io.BufferedReader name=5>
  self.openKicad(kicad_pcb_path)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
****

khaapamyaki commented:
> Cool. Tested with Inkscape 1.0 and the myfirstPCB.svg form the exaples folder.
> It works and the PCB looks fine.
> 
> I get following messages:
> 
> C:\Program Files\Inkscape\lib\python3.8\subprocess.py:942: ResourceWarning: subprocess 24736 is still running
> _warn("subprocess %s is still running" % self.pid,
> ResourceWarning: Enable tracemalloc to get the object allocation traceback
> export.py:488: ResourceWarning: unclosed file <_io.BufferedReader name=4>
> self.openKicad(kicad_pcb_path)
> ResourceWarning: Enable tracemalloc to get the object allocation traceback
> export.py:488: ResourceWarning: unclosed file <_io.BufferedReader name=5>
> self.openKicad(kicad_pcb_path)
> ResourceWarning: Enable tracemalloc to get the object allocation traceback

The error occurs when we activate **Open Kicad after export?**

we could fix the error by adding `.communicate()` to line 488
[python docs](https://docs.python.org/dev/library/subprocess.html#subprocess.Popen.communicate)

```py
self.openKicad(kicad_pcb_path)
```

-> 

```py
self.openKicad(kicad_pcb_path).communicate()
```

fstokke commented:
Yes @daskycodes, I think you are right. Thanks.  

The "subprocess is still running" warning is probably issued because KiCad is left running after the svg2shenzhen script has terminated. 

I will try your suggested fix and update the pull request.

fstokke commented:
I have updated the pull request. Adding .communicate() did the trick, thanks @daskycodes :)

doctormo commented on inkscape/svg2shenzhen/export.py:
```diff
@@ -221,15 +218,15 @@ class Svg2ShenzhenExport(inkex.Effect):
-        self.OptionParser.add_option("--openkicad", action="store", type="inkbool", dest="openkicad", default="true")
-        self.OptionParser.add_option("--autoflatten", action="store", type="inkbool", dest="autoflatten", default="true")
-        self.OptionParser.add_option("--debug", action="store", type="inkbool", dest="debug", default=False)
+        self.arg_parser.add_argument("--path", type=str, dest="path", default="~/", help="")
```
You can use the `add_arguments` method here and clean this up a bit.
```
def add_arguments(self, pars):
    pars.add_arguments("--path", default="~/")
```

You don't have to specify dest if dest is the same as the flag, and type=str is the default.

fstokke replied:
Done

doctormo commented on inkscape/svg2shenzhen/export.py:
```diff
@@ -568,7 +568,7 @@ def openKicad(self, kicad_file_path):
             command = "start %s" % (kicad_file_path)
 
-        return subprocess.Popen(command.encode("utf-8"), shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        return subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
There's an `inkex.commands` module which may (or may not) be helpful, so you don't have to use Popen directly.

fstokke replied:
processExportLayer() method uses Popen to start multiple processes in parallel. I don't think the inkex.commands module supports this. Suggest we keep Popen to avoid introducing multiple methods for starting a process.

doctormo commented on inkscape/svg2shenzhen/export.py:
```diff
@@ -628,32 +628,31 @@ def exportEdgeCut(self, kicad_mod=False):
             for node in self.document.getroot().xpath(nodePath, namespaces=inkex.NSS):
                 d = node.get('d')
-                p = simplepath.parsePath(d)
+                p = inkex.Path(d).to_arrays()
```
This is just `node.path.to_arrays()` no need to convert it yourself.

fstokke replied:
Thanks for the tip. Done!

doctormo commented on inkscape/svg2shenzhen/export.py:
```diff
@@ -628,32 +628,31 @@ def exportEdgeCut(self, kicad_mod=False):
-                            m = simpletransform.parseTransform(t)
-                            trans = simpletransform.composeTransform(layer_m, m)
+                            m = Transform(t).matrix
+                            trans = (Transform(layer_m)*Transform(m)).matrix
```
There's no reason to do `Transform(Transform(t).matrix))` You can just change `m` to `t` and remove line 646

fstokke replied:
Well spotted. Done.

doctormo commented on inkscape/svg2shenzhen/export.py:
```diff
@@ -628,32 +628,31 @@ def exportEdgeCut(self, kicad_mod=False):
             layer_trans = layer.get('transform')
             if layer_trans:
-                layer_m = simpletransform.parseTransform(layer_trans)
+                layer_m = Transform(layer_trans).matrix
```
This can be just `layer.transform.matrix`, it will return a clean transform even if it's empty.

fstokke replied:
I'm on thin ice here. I'm not familiar with the Inkscape APIs and don't know how to test this. 

Where can I find API docs for the layer.transform method?

doctormo replied:
Docs wise transform is a shape level attribute object (like style) example of use https://gitlab.com/inkscape/extensions/-/blob/master/inkex/elements/_base.py#L430

doctormo commented on inkscape/svg2shenzhen/export.py:
```diff
@@ -730,8 +727,8 @@ def flatten_bezier(self):
             for node in layer.xpath(nodePath, namespaces=inkex.NSS):
```
addNS isn't reuquired any more. You can do isinstance(node, inkex.PathElement) there's also an iterator which can just give you all path elements without having to test for them (or use xpath directly)

fstokke replied:
I don't know the APIs well enough to rewrite this.

doctormo commented on inkscape/svg2shenzhen/prepare.py:
```diff
@@ -62,19 +62,19 @@ def setDocumentSquare(self, width, height):
     def createLayer(self, layer_name):
         svg = self.document.xpath('//svg:svg',namespaces=inkex.NSS)[0]
-        layer = inkex.etree.SubElement(svg, 'g')
+        layer = etree.SubElement(svg, 'g')
```
These three lines can be replaced with `layer = svg.add(Layer.new(layer_name))` the API does a lot of this work for you.

fstokke replied:
I'm not familiar with the Inkscape APIs. I'm not comfortable making changes that may have side effects I'm not aware of.

doctormo commented on inkscape/svg2shenzhen/prepare.py:
```diff
@@ -62,19 +62,19 @@ def setDocumentSquare(self, width, height):
 
     def createWhitebg(self):
-        rect = inkex.etree.Element(inkex.addNS('rect','svg'))
+        rect = etree.Element(inkex.addNS('rect','svg'))
```
This can now be `Rectangle.new(0, 0, w, h)` while style can be set with just `rect = {'fill': ...}`

fstokke replied:
Same here

fstokke committed:
Fixed review comments from @doctormo.

fstokke commented:
Thanks for review @doctormo. I have pushed a new commit addressing some of your comments.

I have just followed the migration guide for deprecated APIs and done the bare minimum to get it working with Inkscape 1.0. As you point out some of the code can be improved, but I'm not familiar with the Inkscape APIs and don't want to make changes that might have side effects.

doctormo commented:
@fstokke Completely fair. Thanks for reading my review and fixing some of the issues.

fstokke commented:
@doctormo: I see you are one of the Inkscape API developers (https://inkscape.org/forums/extensions/new-api-for-inkscape-10/), wow! Many thanks for taking your time to do review this.

angelnu commented:
Just to confirm that just by replacing both python files from the fork repository it worked in my Mac. Thanks - waiting for the oficial release.

badgeek commented:
thanks to all who contribute, i will add your credit name to the main page 😸 ❤️
